python-package/lightgbm/basic.py

```diff
@@ -1297,6 +1492,74 @@ def _lazy_init(self, data, label=None, reference=None,
         # set feature names
         return self.set_feature_name(feature_name)
 
+    def __yield_row_from(self, seqs, indices):
+        # type: (List[Sequence], Iterable[int]) -> ...
+        offset = 0
+        seq_id = 0
+        seq = seqs[seq_id]
+        for row_id in indices:
+            assert row_id >= offset, "sample indices are expected to be monotonic"
+            while row_id >= offset + len(seq):
+                offset += len(seq)
+                seq_id += 1
+                seq = seqs[seq_id]
+            id_in_seq = row_id - offset
+            row = seq[int(id_in_seq)]
+            yield row if row.flags['OWNDATA'] else row.copy()
+
+    def __sample(self, seqs, total_nrow):
+        # type: (List[Sequence], int, int) -> (np.ndarray, List[np.ndarray])
+        """Sample data from seqs.
+
+        Mimics behavior in c_api.cpp:LGBM_DatasetCreateFromMats()
+
+        Returns
+        -------
+            sampled_rows, sampled_row_indices
+        """
+        indices = self.create_sample_indices(total_nrow)
+
+        # Select sampled rows, transpose to column order.
+        sampled = [row for row in self.__yield_row_from(seqs, indices)]
+        sampled = np.array(sampled)
+        sampled = sampled.T
+
+        filtered = []
+        filtered_idx = []
+        sampled_row_range = np.arange(len(indices), dtype=np.int32)
+        for col in sampled:
+            col_predicate = (np.abs(col) > ZERO_THRESHOLD) | np.isnan(col)
+            filtered_col = col[col_predicate]
+            filtered_row_idx = sampled_row_range[col_predicate]
+
+            filtered.append(filtered_col)
+            filtered_idx.append(filtered_row_idx)
+
+        return filtered, filtered_idx
+
+    def __init_from_seqs(self, seqs, params_str, ref_dataset):
```

Comment:
Unused parameters `params_str` and `ref_dataset`. It seems that `ref_dataset` is ignored when we constructing from `seqs`. Does that mean this method only supports constructing training dataset? Because for a validation dataset, a reference training dataset is required in Python API.

Reply:
Very much appreciated for making clear on what `ref_dataset` does.
 
>  Does that mean this method only supports constructing training dataset? 
 
Currently, yes. However after your explanation, It seems validation dataset can also be supported via trivial changes to make.

Reply:
Ok. I think now that we've supported creating from `seqs` for training data, we should also support that for validation data. Otherwise, users can get unexpected behavior if they try to create validation dataset from `seqs`. Because the only difference between validation and training datasets in Python API is whether `reference` parameter in `lightgbm.Dataset` is provided.

I'm willing to provide any help.